### ISSUE TITLE
Profiling

### ISSUE THREAD

**Issue by rafcolson** [GoranM/bdx#76]
>The easiest test you can run is to actually time the relevant functions, and see how long they take to process, relative to everything else: scene.update() + renderScene().
>
>You should be able to use com.badlogic.gdx.utils.TimeUtils:
>
>```java
long t = TimeUtils.millis();
functionYouWantToTime();
long tt = TimeUtils.millis();
long delta = tt - t;
```
>
>If there is truly a significant performance benefit, I would love to know, of course, but I think it's something we should discuss in a different issue (we're quite off-topic in this one).

I'll be using this issue for relevant information about profiling. Anyone is welcome to join in.

**Comment by SolarLune** [comment]
That looks cool; I'd like to see a full-on profiling info function or dictionary to assist with finding optimizations. See [this](http://blenderartists.org/forum/showthread.php?355758-BDX-alpha-release-Testers-wanted&p=2801956&viewfull=1#post2801956) page of the BDX BlenderArtists thread.

**Comment by rafcolson** [comment]
I assigned myself, but let this not keep you from joining in. My idea and implementation of a profiler probably won't be exactly what we need, but I'll pull some requests anyway. This way we could discuss it further.

**Comment by rafcolson** [comment]
I've been working on [profiler-scene](https://github.com/rafcolson/bdx/tree/profiler-scene). Everything should work, but no objects are displayed. I think I'm close, though. Any help is appreciated.

I'm not very sure what sizes I should use to instantiate the Camera, is it resolution or units?

**Comment by rafcolson** [comment]
@GoranM Could you please have a look at the API, if my additions regarding the Profiler are correct? Thanks.

**Comment by SolarLune** [comment]
If the profiler runs by default and provides information on each aspect of the game, then it might be a good idea to put that somewhere on those pages. Maybe under the main heading of the Profiler, a bit could be written concerning how to use it.

For example, to just call percents() to get the hash map of each "aspect" of the engine running and the corresponding percentages, what each profiled stat corresponds to, what a is good spread, so to speak (i.e. if you have more on `Outside`, it means your game isn't using all of the computer's resources, and so is running at an excess, which is good), and so forth.

**Comment by GoranM** [comment]
I'm not sure if there should be a link to the Profiler page under com.nilunder.bdx.utils: It's not really a class for the user to manage, or use directly - It's basically a system object with a few relevant methods that developers can actually use (start, stop, percents).

I think the profiler API page should link only from the Bdx page, and I think it should only document methods intended for developer use (instead of everything that's on the class).

Also, there's no need to document inherited methods, unless their behavior was changed with an override - The fact that it's a hashmap is implied, since you note that it extends LinkedHashMap (btw: It should be LinkedHashMap\<String, Long\>), so there's no need to document "get" and "entrySet".

You can just say: This is a hash map that contains execution times (in milliseconds) for various subsystems, and people will know what you're talking about, just from the type signature.

Although, it doesn't seem like you can just "get" the time when processing logic, because you're still in the process of calculating the deltas at that point - so, for now, only percents is valid.

**Comment by rafcolson** [comment]
Okay, thanks guys, I'll look into it later this day.

**Comment by GoranM** [comment]
Well, I think the developer could just use Bdx.profiler:

```java
Bdx.profiler.start();
// code
Bdx.profiler.stop("code");
```

The "code" category would then show up in the profiler display (when we make one). With that, I don't think the developer would ever need to make their own instance of Profiler.

**Comment by rafcolson** [comment]
Okay, I'm beginning to understand.

But wouldn't this mess things up? For example, when Bdx.profiler is already timing ```Scene.runObjectLogic()``` which contains ```GameObject.main()```. Or ```"outside"``` which contains ```GameObject.init()``` (I think).

Isn't Bdx.profiler suppose to time "everything"? So any additional timing should be on another 'layer' sort of speak.

I was thinking, would it really be necessary for the user to have manual access to Bdx.profiler anyway. If he wants to see what's going on, wouldn't the profiler overlay scene be enough. Else, I think Bdx.profilerPercents() would be better, so the user has no access to Bdx.profiler whatsoever.

What are your thoughts?

**Comment by GoranM** [comment]
Well, yea, it's not compatible with the current profiler design, but I'm thinking that maybe we should redesign to support that.

> Isn't Bdx.profiler suppose to time "everything"? So any additional timing should be on another 'layer' sort of speak.

Game code is part of "everything", and I don't think it would be all that difficult to incorporate a custom category, which would just show up as a percentage of overall game time.

I don't really know ... I'll have to think about this for a bit.

For now, what you have is fine - Just remove documentation for inherited methods, and leave everything else as it was.

**Comment by rafcolson** [comment]
>Game code is part of "everything", and I don't think it would be all that difficult to incorporate a custom category, which would just show up as a percentage of overall game time.

If this option would be added to Bdx.profiler (so without having to make a separate instance of Profiler) then next to ```deltaTime``` values also ```startTime``` values for every category. Also the current (rather easy) workflow wouldn't be possible, to just add one ```stop(categoryName)``` after the other. Then you should always have to 'define' the start, so you wouldn't be having ```start()``` but ```start(categoryName)``` instead. Actually, this is how I had it in the beginning (in order to support adding categories in the same profiler).

Personally I like the current workflow because:
1. There're no risks of overriding categories, for example, you couldn't do ```Bdx.profiler.stop("graphics")``` if you're adding categories to Bdx.profiler.
2. Less code (continious ```stop()```'s and no need to fill in a category for ```start```)

**Comment by GoranM** [comment]
The current workflow doesn't really seem that easy to me. I mean, you have to create a new instance whenever you want to time something, and the results are not relative to overall simulation time, so the percents don't really mean much.

Also, display is not automatic - You would either have to merge the data into a running profile display, or you have to make a unique display just for the given profiler.

> wouldn't be possible, to just add one stop(categoryName) after the other

If stop("category") is called, and there is no previous start time for that category, I think you can simply use the time of the last stop call as your start time.

```java
start("blah")
// code
stop("blah")
// more code
stop("meh") // startTime = lastStopTime
```

> There're also no risks of overriding categories, for example, you couldn't do Bdx.profiler.stop("graphics")

We could simply throw an exception if someone tries to use reserved category names.

I'm not saying I'm absolutely sure that this would be better, but it's something to think about.

**Comment by rafcolson** [comment]
What I mean is this:
```java
Bdx.profiler.stop("logic"); // this should be the starting point for "visuals" in Scene.java
...
...
Bdx.profiler.start("myCategory"); // this will overwrite that starting point
...
Bdx.profiler.stop("myCategory"); // this will too
...
Bdx.profiler.stop("mySecondCategory"); // this too.
...
...
Bdx.profiler.stop("visuals"); // so this will return the wrong delta
```
Even if you would store another variable ```lastStopTime```, also this would be overwritten. The only solution is to store the ```startTime``` for each category. So you can't have continuous ```stop()```'s anymore.

Then you could do like this without any problems:
```java
Bdx.profiler.stop("logic");
Bdx.profiler.start("visuals");
...
...
Bdx.profiler.start("myCategory");
...
Bdx.profiler.stop("myCategory");
...
Bdx.profiler.stop("mySecondCategory");
...
...
Bdx.profiler.stop("visuals");
```
Or maybe I don't understand at all what you mean. In that case, I'll leave this completely up to you. Just for the sake of not wasting too much time on this going back and forth.

So let me know if I could redesign it to the latter.

**Comment by GoranM** [comment]
When you call start, you would have to specify the category, yes, but I don't think that would prevent you from having multiple stop calls without a start:

```java
Profiler p = Bdx.profiler;

p.start("logic");
...
...
p.start("myCategory");
...
p.stop("myCategory");
...
p.stop("mySecondCategory");
...
...
p.stop("logic");
```

So, when stop is called, it could check:

```java
stop = TimeUtils.millis();
start = startTimes.get(category);
if (start == null)
    start = lastStopTime;
lastStopTime = stop;
```

And that would be the same as if you called p.start("mySecondCategory") right after p.stop("myCategory"), right?

**Comment by rafcolson** [comment]
Yes, you're right. That would do it. I'll update the profiler shortly (within a couple of days). Thanks.

**Comment by rafcolson** [comment]
@SolarLune 

You wrote about the profiler scene:
>So the scene would be packaged in the scenes resource directory, along with the actual game scenes that BDX exports?
I don't know much about what you are saying. Could you please elaborate? How would I package the scene? And in which resource directory? And where can I find the code that imports scene data?

I think I have most of the code done. I only have to get the data from the profiler blend file. But I don't know how.

**Comment by SolarLune** [comment]
When you run BDX, it exports the scenes to files in the /android/assets/bdx/scenes/ folder. I'd imagine an easy way to go about this would be to write your profiler scene as you want to have it work, and then put the resulting profiler scene file in that folder, and the profiler class file in the source folder (next to BdxApp.java). For developers who want to improve the profiler scene further, you can put the profiler.blend file somewhere (maybe upload it here or something). 

Maybe not, though; Goran would know exactly how he wants things done.

**Comment by GoranM** [comment]
I would want the simplest possible implementation, and the more I think about it, the more it seems like we should simply include the profiler scene in the default .blend file.

That will export via the normal process, and then the existing Profiler object can handle updating that scene - After Profiler updates that scene, we render it on top of everything.

It's simple, while serving all our profiling needs.

**Comment by rafcolson** [comment]
Okay, thanks.

**Comment by rafcolson** [comment]
I still don't understand. Are the objects for the in the default .blend file exported only once? So if I add a new scene with all the object for the profiler to "../bdx/gen/game.blend", how do I access those?

**Comment by GoranM** [comment]
Assuming your scene is named "Profiler":

```java
profilerScene = new Scene("Profiler");
GameObject whatever = profilerScene.objects.get("Whatever");
whatever.applyForce(0, 0, 300);
profilerScene.update();
```

**Comment by rafcolson** [comment]
If I understand correctly, then this profiler scene would have to be part of the .blend in the project directory (.../project_name/project_directory/blender/game.blend), right?

But wouldn't we want the profiler to work regardless of which scenes the user has in his game.blend? Is pretty inconvenient, isn't it? Or am I missing something?

**Comment by GoranM** [comment]
> It's pretty inconvenient, isn't it?

I don't think so: If it's in the template .blend, the new project .blend will automatically contain the same scene - The user doesn't have to do anything manually.

**Comment by rafcolson** [comment]
I think it's a mistake to make something that's part of the engine a part of the user's project.

It's just one camera, one plane and one text object I need for the profiler scene. Far more straightforward to create those with gdx. And it wouldn't be an exception to the rule, as the Scene class if full of those dependencies.

So I'm sorry to say that I'm to uncomfortable with it. I'm imagining users wanting to use the profiler, even though they've been working on their projects without, and then questioning why they actually have to have a scene on top of their project in order for it to work.

**Comment by GoranM** [comment]
> It's just one camera, one plane and one text object. Far more straightforward to create those with gdx.

I don't see how it would be more straightforward to write additional code for something that's already done by exporter, and the existing scene system.

> I'm imagining users wanting to use the profiler, even though they've been working on their projects without, and then questioning why they actually have to have a scene on top of their project in order for it to work.

We have virtually no users at this point, and that's unlikely to change in the foreseeable future, so there's really nothing to worry about in that regard.

Simply adding the profiler scene to the default .blend, and manipulating that scene as I demonstrated above, seems like the easiest/simplest thing we could do to implement this feature, because it builds on top of existing systems.

I agree that packaging the profiler scene in the project .blend is not ideal, but I think it's best to do that as the first step - After that's in master, and working, we can focus on how best to package that scene externally.

**Comment by rafcolson** [comment]
Well, the problem with this temporal solution is that it needs temporal code for scene management too. To keep the profiler scene always on top and free from filters, you'd have to corrupt Bdx.main() and Bdx.ArrayListScenes with these temporal solutions. I'd rather spend time on doing it right from the beginning.

So if you don't like this scene to be a single java script, then let's make it a separate .blend altogether.

Anyway, it's over my head right now. There's still a lot of the code I don't quite grasp.

**Comment by GoranM** [comment]
The only temporary part is in how we package the scene (in the .blend vs externally) - the code that loads/manages that scene won't have to change, so that is "doing it right from the beginning".

> Anyway, it's over my head right now. There's still a lot of the code I don't quite grasp, especially the code that exports the Blender scenes.

You don't have to know any details about the export process in order to do what I suggested. You can just load the scene as I demonstrated in the existing Profiler object. The you can update it there, and then finally draw it after all the other scenes have been rendered.

There's nothing to "corrupt", because the scene is never going to be in Bdx.scenes - it's just a scene that you manage yourself, in the profiler, outside the normal update loop.

**Comment by SolarLune** [comment]
I'd rather the scene not be something the user can delete or mangle otherwise as well, but if one does, the game won't compile, so it's not a huge deal, I think.

An alternative approach would be to put the blend somewhere (like say, the `blender\bdx\gen` folder) for editing for developers, and just put the exported profiler scene file in the scenes folder as I mentioned above. The Profiler blend file itself wouldn't exist in a place that the user could reasonably reach and mess with it. BDX could, on run, export the Profiler scene as an extra step if the exported Profiler scene doesn't exist yet.

**Comment by GoranM** [comment]
> I'd rather the scene not be something the user can delete or mangle otherwise as well, but if one does, the game won't compile, so it's not a huge deal, I think.

It would compile, but it would crash immediately because `new Scene("Profiler")` would fail to load a non-existent `Profiler.bdx`.

As for how we would actually package the profiler scene: Yeah, there are many different ways to do it, and that's something we need to think about, but the underlying profiler systems, that would manage the scene, will be the same regardless of how we package and load the scene data, so it makes sense to do the most straightforward thing first, and simply include the profiler scene in the default .blend.

> just put the exported profiler scene file in the scenes folder

Well, it's a bit more involved, because .bdx files don't actually include things like textures and font data - they simply reference those resources, which we would need to place in their proper directories.

**Comment by rafcolson** [comment]
PR submitted.

**Comment by rafcolson** [comment]
@GoranM I'm glad you like it. But I forgot to push the last changes to the commit. The only change was the output of function ```Bdx.profiler.stop("myFunction")```. Currently, as it is merged, this will return the ```deltaTime``` in Long nanoseconds. But shouldn't I have to change this into Float milliseconds? I think this is an easier format for the user to work with. So let me know if you think it's a good idea.

**Comment by GoranM** [comment]
PR merged.

The API looks fine, but I would just change:

> This class manages execution times for various subsystems.

To:

> This class measures execution times for various subsystems.

**Comment by rafcolson** [comment]
So now, if you run the game once (so it builds ```__Profiler.bdx```) and then remove ```__Profiler``` scene, it will still work because ```__Profiler.bdx``` will remain in  ```assets/scenes/```.

>Yeah, that's how it works - As long as the bdx file is in the right place, along with all the resources it requires, the scene itself doesn't have to be present in the .blend
>
>I don't think it should be too difficult to leverage that, so we don't have to include the profiler scene in the .blend, but this is ok for now.

Any ideas?

**Comment by rafcolson** [comment]
So the code for this should do two things:
- save the profiler scene to *.bdx
- remove the profiler scene from the default blend

So I think all that's required is to run this code directly after the project is created:
```python
import os

# Export profiler scene to *.bdx:
j = os.path.join
asset_dir = j(proot, "android", "assets", "bdx")
profilerScene = bpy.data.scenes["__Profiler"]
file_name = profilerScene.name + ".bdx"
file_path = j(asset_dir, "scenes", file_name)
bpy.ops.export_scene.bdx(filepath=file_path, scene_name=profilerScene.name, exprun=True)

# Remove profiler scene from default blend:
bpy.data.scenes.remove("__Profiler")
```
Only thing I'm doubtful about is the ```exprun=True``` argument assignment in the ```export_scene.bdx()``` call. I couldn't find ```bpy.export_scene.bdx()``` anywhere in the database, so I wonder what ```exprun=True``` does. Should it be omitted or set to ```False``` instead?

Thanks.

**Comment by GoranM** [comment]
`bpy.ops.export_scene.bdx` is an operator created by blender, upon registration, from [the id provided](https://github.com/GoranM/bdx/blob/master/blender/bdx/exporter.py#L685) by the ExportBdx class.

It's basically a call into `exporter.export`, which takes all those arguments. The `exprun` argument is there to specify that this operation is in the bdx "export and run" context, instead of just scene export, so that additional files (instantiators, fonts, etc) would be generated.

If you searched the code base for "exprun", you would have probably figured this out on your own.

Anyway:

It's true that __Profiler.bdx exists after initial creation, even after the scene is removed from the .blend, but if you were to update BDX, and simply move the .blend to a freshly generated project, you would have problems, because the scene wouldn't exist, and the exported .bdx file wouldn't exist (same problem with packed BDX projects). It may also prove necessary to remove all the scene files prior to export, [as we currently do for instantiator files](https://github.com/GoranM/bdx/blob/master/blender/bdx/ops/exprun.py#L22).

With that in mind, I would prefer if we could temporarily import the profiler scene from the default .blend (assuming that's possible), just for the export process.

That should work for all cases.

**Comment by rafcolson** [comment]
>If you searched the code base for "exprun", you would have probably figured this out on your own.

Yes, I probably would have figured out that it is an operator. But it's still hard for me to figure out the workings of it just by reading the code. I don't think I quite understand yet, but I get a better idea now. So thanks.

>It's true that __Profiler.bdx exists after initial creation, even after the scene is removed from the .blend, but if you were to update BDX, and simply move the .blend to a freshly generated project, you would have problems, because the scene wouldn't exist, and the exported .bdx file wouldn't exist (same problem with packed BDX projects). It may also prove necessary to remove all the scene files prior to export, as we currently do for instantiator files.
>
>With that in mind, I would prefer if we could temporarily import the profiler scene from the default .blend (assuming that's possible), just for the export process.
>
>That should work for all cases.

Right, so if I understand correctly, I would have to get the scene from the ```game.blend``` in the addons folder (```addons/bdx/gen/game.blend```), convert this temporarily to a *.bdx and store it somewhere for later use. So with every BDX update it should export to *.bdx only once when the project is run for the first time (within that BDX version).

So the code should be in ```exporter.py```, correct?

I'll try some things now and come back to it later.

>It may also prove necessary to remove all the scene files prior to export, as we currently do for instantiator files.

I'll leave this to you as I have problems understanding the mechanics behind this. Maybe in the future I  will understand it better. Sorry about that.

**Comment by GoranM** [comment]
> I don't think I quite understand yet

Fundamentally, the BDX blender addon is just an exporter, which can convert a blender scene into a bdx scene file. The ability to "export and run" (and to generate all the files required for that to work) is extra functionality, which is typically not something that exporters provide.

So, the "exprun" argument is there to specify when we want that extra functionality, instead of simply exporting the scene to some arbitrary directory.

You can look at [how the boolean is actually used](https://github.com/GoranM/bdx/blob/master/blender/bdx/exporter.py#L652), to understand what it enables, in addition to a pure export.

> Right, so if I understand correctly, I would have to get the scene from the game.blend in the addons folder (addons/bdx/gen/game.blend), convert this to a *.bdx and store it somewhere for later use. So with every BDX update it should export to *.bdx only once when the project is run for the first time (within that BDX version).

Before export, you would add the profiler scene from the default game.blend, into the current project game.blend. After export, you would remove it.

> So the code should be in exporter.py, correct?

There's no need to modify the exporter, because you're not fundamentally changing how things are exported - You're just including an additional scene, temporarily.

You could probably just do something like this, in exprun.py:

```python
        # add profiler scene to the .blend

        # Export scenes:
        # ...

        # remove profiler scene from the .blend
```

**Comment by rafcolson** [comment]
PR submitted. Now the profiler scene is internal, also if you would replace the ```game.blend``` immediately after project creation (so without running BDX).

**Comment by rafcolson** [comment]
@SolarLune 
@GoranM 

So to the next stage. I would like to add two things:

- ```avgTicRate```: an average over 60 tics (or 1 second)
- an open mono spaced font (such as [Liberation Mono Bold](http://en.wikipedia.org/wiki/Liberation_fonts) or [Inconsalata](http://www.levien.com/type/myfonts/inconsolata.html) or something you think is best), and use this for the text in the profiler scene. This would remove at least 10 text objects from the scene and should give a nicely stacked table.

Let me know if there are any suggestions.

**Comment by GoranM** [comment]
In addition to monospace, it would probably be a good idea to use a square font; I think that would be readable at any size.

As for the background: You could simply have a semi-transparent black plane behind the text.

> Any idea what kind of font the BGE is using?

No idea, but we don't have to do everything exactly like it's done in the BGE - You can just make the bars from pipe characters: `|||||||||||`

**Comment by rafcolson** [comment]
Is there a type of font you have in mind, perhaps? Or something that comes close? I'll make the font from scratch so it can be part of BDX.

**Comment by GoranM** [comment]
I don't really have a preference, as long as it's a square, monospaced font.

I don't know if there's any utility in "average tick rate", because it's locked to the framerate, but if you want to see that (for whatever reason), feel free to add it.

**Comment by rafcolson** [comment]
PR submitted. Let me know if any changes has to be made. (Hope you like the font).

**Comment by GoranM** [comment]
:+1:

**Comment by SolarLune** [comment]
Hey, thanks a ton for this - it's really nice and well-done. I think the profiler panel as a whole might be a tad big, though?

**Comment by rafcolson** [comment]
I'm happy you like it. Yes, it's a bit to large, isn't it? @GoranM what do you think?

Anyway, I can make it smaller as I'm also changing the "D" of the font. I didn't notice that it looks to much like an "O".

**Comment by SolarLune** [comment]
As a side-note, for the template blend file, it might be a good idea to enable the grid again, in the Properties panel in the 3D view (N-key) and take out the Video Editing screen. Also, if there's still a "game.blend1", it can be removed.

**Comment by rafcolson** [comment]
Okay, sounds good. I'll make those changes in a separate PR after I've made changes to the profiler. But until then, feel free to make those changes yourself. Thanks for the feedback.

**Comment by SolarLune** [comment]
Oh, also, maybe we could adjust the default resolution to something more "usual"? I don't remember the  default, but something 16:9 would be good, I think.

**Comment by GoranM** [comment]
The size looks fine to me - I mean, when debugging, I would want it to be big, but that's just my personal preference. In general, I'm ok with whatever you decide to do, relating to size.

> Also, if there's still a "game.blend1", it can be removed from the repo.

There is no "game.blend1" in the repo.

**Comment by rafcolson** [comment]
>The size looks fine to me - I mean, when debugging, I would want it to be big, but that's just my personal preference. In general, I'm ok with whatever you decide to do, relating to size.

I could add a "size" attribute, so the profiler camera's field of view can be adjusted. Then in the BDX panel we could add a check box "Show Profiler" and another check box "Small" next to it (which will become active when the "Show Profiler" box is checked). This way we can meet the personal preferences of the user.

@SolarLune Yes, I think "800 x 450" should be good. Or "1280 x 720", "1024 x 576", "640 x 360", ... any of those.

**Comment by rafcolson** [comment]
PR submitted for profiler improvements (changes nothing to the "game.blend"). Font adjustments have to wait for another time.

**Comment by rafcolson** [comment]
@GoranM PR submitted.

**Comment by rafcolson** [comment]
Well, after that bug fix I think I can close this now. It was quite a run. Thanks guys for all your help.

There's only one thing that probably could be changed, but it's not such a big deal. The material of the semi-transparent background gets in ```game.blend``` when it runs for the first time. But once ```__Profiler.bdx``` is compiled, the next time(s) you'll run ```game.blend``` it will be gone.
